pkg/cmd/cli/cmd/logs.go

```diff
@@ -0,0 +1,176 @@
+	cmd.Flags().BoolVar(&o.Interactive, "interactive", o.Interactive, "If true, prompt the user for input when required. Default true.")
+	cmd.Flags().BoolVarP(&o.Previous, "previous", "p", o.Previous, "If true, print the logs for the previous instance of the container in a pod if it exists.")
+	cmd.Flags().StringVarP(&o.ContainerName, "container", "c", o.ContainerName, "Container name")
+	return cmd
+}
+
+func (o *OpenShiftLogsOptions) Complete(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args []string) error {
+	switch len(args) {
+	case 0:
+		return cmdutil.UsageError(cmd, "RESOURCE is required for log")
+
+	case 1:
+		o.ResourceString = args[0]
+	case 2:
+		o.ResourceString = args[0]
+		o.ContainerName = args[1]
+
+	default:
+		return cmdutil.UsageError(cmd, "log RESOURCE [CONTAINER]")
+	}
+
+	var err error
+	o.Namespace, _, err = f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	o.OriginClient, o.KubeClient, err = f.Clients()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *OpenShiftLogsOptions) Validate(cmd *cobra.Command) error {
+	return nil
+}
+
+func (o *OpenShiftLogsOptions) RunLog() error {
+	kLogsOptions := &kcmd.LogsOptions{
+		Client: o.KubeClient,
+
+		PodNamespace:  o.Namespace,
+		ContainerName: o.ContainerName,
+		Follow:        o.Follow,
+		Interactive:   o.Interactive,
+		Previous:      o.Previous,
+		Out:           o.Out,
+	}
+
+	resourceType := "pod"
+	resourceName := o.ResourceString
+	tokens := strings.SplitN(o.ResourceString, "/", 2)
+	if len(tokens) == 2 {
+		resourceType = tokens[0]
+		resourceName = tokens[1]
+	}
+	resourceType = strings.ToLower(resourceType)
+
+	// if we're requesting a pod, delegate directly to kubectl logs
+	if (tokens[0] == "pods") || (tokens[0] == "pod") || len(o.ContainerName) > 0 {
+		kLogsOptions.PodName = resourceName
+		return kLogsOptions.RunLog()
+	}
+
+	switch resourceType {
+	case "bc", "buildconfig", "buildconfigs":
+		buildsForBCSelector := labels.SelectorFromSet(map[string]string{buildapi.BuildConfigLabel: resourceName})
+		builds, err := o.OriginClient.Builds(o.Namespace).List(buildsForBCSelector, fields.Everything())
+		if err != nil {
+			return err
+		}
+		if len(builds.Items) == 0 {
+			return fmt.Errorf("no builds found for %v", o.ResourceString)
+		}
+
+		sort.Sort(sort.Reverse(buildapi.BuildSliceByCreationTimestamp(builds.Items)))
+
+		return o.runLogsForBuild(&builds.Items[0])
+
+	case "build", "builds":
+		build, err := o.OriginClient.Builds(o.Namespace).Get(resourceName)
+		if err != nil {
+			return err
+		}
+		return o.runLogsForBuild(build)
+
+	default:
+		return fmt.Errorf("unknown resource type %v", resourceType)
+	}
+}
+
+func (o *OpenShiftLogsOptions) runLogsForBuild(build *buildapi.Build) error {
+	opts := buildapi.BuildLogOptions{
+		Follow: o.Follow,
+		NoWait: false,
+	}
+
+	readCloser, err := o.OriginClient.BuildLogs(build.Namespace).Get(build.Name, opts).Stream()
+	if err != nil {
+		return err
+	}
+	defer readCloser.Close()
```

Comment:
As a note - deferring both unlock and close are idiomatic - arguments must
be made for the opposite, and by default we should be structuring code
(within reason) so that it easily handles using defer.

On Fri, Oct 9, 2015 at 7:44 AM, David Eads <notifications@github.com> wrote:

> In pkg/cmd/cli/cmd/logs.go
> <https://github.com/openshift/origin/pull/4997#discussion_r41621702>:
>
> > +	default:
> > +		return fmt.Errorf("unknown resource type %v", resourceType)
> > +	}
> > +}
> > +
> > +func (o *OpenShiftLogsOptions) runLogsForBuild(build *buildapi.Build) error {
> > +	opts := buildapi.BuildLogOptions{
> > +		Follow: o.Follow,
> > +		NoWait: false,
> > +	}
> > +
> > +	readCloser, err := o.OriginClient.BuildLogs(build.Namespace).Get(build.Name, opts).Stream()
> > +	if err != nil {
> > +		return err
> > +	}
> > +	defer readCloser.Close()
>
> Do we really need to defer here?
>
> Yes. Its idiomatic and it ensures that the stream will be closed even if
> something weird happens in the Copy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4997/files#r41621702>.
>